### ISSUE TITLE
Collect data from model environment and agent simulatenously.

### ISSUE THREAD

**Issue by itsdfish** [JuliaDynamics/Agents.jl#188]
Currently, the data collection feature is limited to information about the agent or the model. It would be convenient to have the ability to collect information from both. In addition, it might be useful to collect variables that are computed from other variables (although this might not be easy to do in a general way).

In the meantime, I have found that it is possible to pass a dataframe and collect the information manually like so:

`step!(model, (x,y)->agent_step!(x, y, df), model_step!, 150)`

**Comment by Libbum** [comment]
Yep, this is certainly a limitation of our currently implementation. (See #151 also). I'm currently looking at scenarios that better satisfy all user needs in this matter and we'll be reworking the data collection methods for v3.

Will begin discussions over at #187.

**Comment by Datseris** [comment]
When data are aggregated, you can still get model variables, as well as "variables from variables" (since you can pass any arbitrary function). E.g. use `Dict(:model =>[...], :wealth => [...])`

I assume that what you really try to say @itsdfish is that you want to collect model data at every step using the `properties::Vector{Symbol}` case. In this case, what is the `id` field of the model? (The only thing that distinguishes agents in this part of the data collection is the ID. What is it for the model?)

---

> In addition, it might be useful to collect variables that are computed from other variables (although this might not be easy to do in a general way).

Please explain this in more detail.

**Comment by itsdfish** [comment]
Perhaps the capabilities for these features already exist. I'll try to give more concrete examples of each use case just to clarify.

Suppose that there is a food resource in the environment that replenishes at some rate as long as at least one unit exists. In such a model, it would be useful to track the quantity of food available for each agent's interaction. The resulting data for two agents on step 1 might look like this:

```
2×3 DataFrame
│ Row │ id    │ food_units │ step  │
│     │ Int64 │ Int64      │ Int64 │
├─────┼───────┼────────────┼───────┤
│ 1   │ 1     │ 100        │ 1     │
│ 2   │ 2     │ 92         │ 1     │

```

As you noted, in some cases it is possible to compute variables from other variables in post processing. However, there are cases where that is difficult. Suppose I have a model in which agents interact at each step. How would I collect the id of the agents in each interaction and the result of that interaction? Perhaps based on some combination of variables, the interaction could be classified as good or bad. 

```
2×4 DataFrame
│ Row │ id    │ other_id │ step  │ interaction │
│     │ Int64 │ Int64    │ Int64 │ Symbol      │
├─────┼───────┼──────────┼───────┼─────────────┤
│ 1   │ 1     │ 3        │ 1     │ good        │
│ 2   │ 2     │ 1        │ 1     │ bad         │

```
What is important is that some variables (e.g. other_id, interaction) are not properties of the model or agent arguments in `agent_step!`. I suppose that I could add other_id and interaction as fields of the agents. However, there are downsides to this approach. First, it might make the model less readable to assign properties to agents that they do not "own" for lack of better word. Second, it could become tedious to enumerate a large list of "interaction" variables.

If this is currently possible, it might be good to have some simple examples in the documentation. I could add that once I understand how to do it.

**Comment by Datseris** [comment]
```
2×3 DataFrame
│ Row │ id    │ food_units │ step  │
│     │ Int64 │ Int64      │ Int64 │
├─────┼───────┼────────────┼───────┤
│ 1   │ 1     │ 100        │ 1     │
│ 2   │ 2     │ 92         │ 1     │
```
This tells me that `food_units` is a property of each agent, since it is specified by `id`, and thus this is already possible. If it is a property of the model instead, how did you create the `id` column?

The second example you mention with `other_id` is not possible without making the `other_id`  a field the agent. Unfortunately, I don't know how to make this possible either (programming-wise).

**Comment by Libbum** [comment]
> How would I collect the id of the agents in each interaction and the result of that interaction? 

I think this example may be a bit too contrived to give an answer here just yet. I follow the context, but without understanding how the interaction became good or bad for each agent pair, it seems the only logical solution would be (as @Datseris said), include this tracking somewhere. Personally, I'd say this is something like `Dict((id, id), bool)` in the model.

**Comment by itsdfish** [comment]
I don't think food_units is necessarily a property of the agent in my example. Suppose agents have access to a common food source:

```
mutable struct Environment
    food_units::Int
end

environment = Environment(100)
model = ABM(Forager, properties=environment)
```

In my example, I would like to track the food_units available in the environment for each agent at each step. I could add food_units to the agent for the purpose of recording data, but it seems odd conceptually. 

I suppose my other example is contrived in some sense. I left out some details because I was interested in a general method for creating a variable not in the model or agent. This might be too difficult to do in a general way. 

The only solution I can think of is the following:

```
df = DataFrame(...)
step!(model, (x,y)->agent_step!(x, y, df), model_step!, n_steps)
```
And manually add the variables I want to the dataframe:

```
function agent_step!(agent, model, df)
    # do some stuff
     # append variables to dataframe
end
```

**Comment by Datseris** [comment]
I am still confused about whether `Environment` is a _model parameter_ or an _agent property_. Clearly, from the code snippet you provide, it is a model property. Yet in the words you write below the code, it is an agent property, as if each agent has their own environment and food resources.

Which one is it? 

Or, maybe, what you care is the actual value of `Environment` at each time `agent_step!` is called for a given agent?

---

The manual overwrite you provide makes sense to me and should be the way to go. Please read at my comment here : https://github.com/JuliaDynamics/Agents.jl/issues/187#issuecomment-606455516  (and the preceeding comment) . I think making such a modular interface will make it easier for users with advanced data collection needs. Let me know if what I sketch there is helpful for you.

**Comment by Datseris** [comment]
When we have a better `step!`, we should put this idea in the docs: 

```
df = DataFrame(...)
step!(model, (x,y)->agent_step!(x, y, df), model_step!, n_steps)

function agent_step!(agent, model, df)
    # do some stuff
     # append variables to dataframe
end
```
i think its simple.

**Comment by itsdfish** [comment]
> Or, maybe, what you care is the actual value of `Environment` at each time `agent_step!` is called for a given agent?

Yeah. It is a model property that is altered by each agent's decision. That seems to describe my situation. The simplest example I can think of is something from game theory in which agents fish from a public pond. 

The new API looks more flexible and should cover all use cases. The main downside is that it requires the user to manually reproduce common patterns that are currently handled automatically in step. Aside from that, I think it looks good. 

Yes. Please add my example to the docs if you find it helpful. Fortunately, Julia makes it easy to hack into things when you want to do something that falls outside of a package's API.

**Comment by Datseris** [comment]
> The new API looks more flexible and should cover all use cases. The main downside is that it requires the user to manually reproduce common patterns that are currently handled automatically in step. Aside from that, I think it looks good.

This is not an issue. All currently existing API will be offered as high level functions once we refactor the old `step!`.

**Comment by itsdfish** [comment]
Awesome. In that case, I think the new setup will be convenient and flexible. Nice work!

**Comment by Datseris** [comment]
Heh, the work hasn't been done yet, and this part will take **a lot** of work!
